### PR TITLE
Test/fff

### LINKED ISSUE #5: task: fix tests structure
### Description:

Right now all tests runs with a single main, which defeats the purpose of separating them in single file for module. As we recently discovered how to properly structure them in folders, they can be updated.

Also **FFF** needs to be added to each test as it's been decided in the last meeting.

### Implementation steps:

1. Remove headers (not needed anymore)
3. Move tests in subfolders of `test/`
4. Create a `main` for each file
5. Add FFF to remove manually implemented mocking (copy the `fff.h` file to some place out of the way, e.g. `test/include/`)
6. Properly check if some tests are missing (e.g. some edge case...)
7. Modify `platformio.ini` accordingly
8. Try them

### Dependencies:

- [FFF](https://github.com/meekrosoft/fff)

### Additional notes:

None

### PR DESCRIPTION
**Pull request by MattWaX**
# Purpose
Clearly state the purpose of this pull request.

Refactor old test structure.

# Overview
Summarize the key changes introduced in this PR.

- Ditch old mockup functions in favor for `fff.h` model
- Make the test more modular to apply filters

# Guidance
Provide guidance to reviewers about the order in which to review the files.
Recommend where to start and how to proceed with the review.

The changed code is in the `test` subfolders:
```
test
├── include
├── test_feedback
│   └── test_feedback.c
└── test_indicator
    └── test_indicator.c
```

### CODE REVIEW

File: test/test_feedback/test_feedback.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    feedback_init(fake_read_feedback);
+}
+
+/*!
+ * \defgroup feedback Test cases for feedback module functions
+ * \{
+ */
+
+void test_feedback_get_state_call_count(void) {
```
Comment (Bridiro) [inline test/test_feedback/test_feedback.c:27]: this name does not make sense to me. In this test you're testing `feedback_update_state()`, so why does the name of test tell me you're testing `get_state`?
The name of the test should explicitly tell you what you're doing, nearly like a full English sentence (e.g. `test_feedback_update_state_call_count`)

File: test/test_feedback/test_feedback.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+void test_feedback_get_state(void) {
+    // Run the update cycle
+    feedback_update_state();
+
+    // Verify States via getter
+    TEST_ASSERT_EQUAL_INT_MESSAGE(
+        FEEDBACK_STATUS_HIGH,
+        feedback_get_state(FEEDBACK_NAME_STEERING_WHEEL),
```
Comment (Bridiro) [inline test/test_feedback/test_feedback.c:45]: you can't use a function that you didn't even tested to check the correctness of another one.
You should `extern` the handler of the module (go check other repositories for examples on how to, e.g. steering-wheel-sw, ecu-sw...) and access directly it.

File: test/test_feedback/test_feedback.c
Comment (Bridiro) [inline test/test_feedback/test_feedback.c:1]: You're not even testing `feedback_get_state` and `feedback_init`. Tests should be exhaustive, on every function and on meaningful cases for each.

File: test/test_indicator/test_indicator.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    indicators_set(true, INDICATORS_NAME_TS_OFF);
+    TEST_ASSERT_TRUE_MESSAGE(
+    indicators_global_handler.state[INDICATORS_NAME_TS_OFF],
+    "TS_OFF indicator should be ON");
+
+    indicators_set(true, INDICATORS_NAME_TSAL);
+    TEST_ASSERT_TRUE_MESSAGE(
+    indicators_global_handler.state[INDICATORS_NAME_TSAL],
+    "TSAL indicator should be ON");
+}
```
Comment (Bridiro) [inline test/test_indicator/test_indicator.c:58]: split this in 4 separate tests

File: test/test_indicator/test_indicator.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        is_indicator_initialized,
+        "Double initialization should fail");
+}
+
+void test_indicator_initialization_state(void) {
+    // After initialization, all indicators should be off
+    for (int i = 0; i < INDICATORS_NAME_COUNT; i++) {
+        TEST_ASSERT_FALSE_MESSAGE(
+            indicators_global_handler.state[i],
+            "Indicator should be off after init");
```
Comment (Bridiro) [inline test/test_indicator/test_indicator.c:34]: cycles should be avoided in tests. You can use test functions or macro such as `TEST_ASSERT_EQUAL_<type>_ARRAY(arr1, arr2, len)`

File: test/test_indicator/test_indicator.c
Comment (Bridiro) [inline test/test_indicator/test_indicator.c:1]: you're not testing the behavior of `indicators_init` with valid inputs

**Review by Bridiro — CHANGES_REQUESTED** [review]
Needs more tests as it's not exhaustive enough.

File: test/include/unity_config.h
Comment (Tonidotpy) [inline test/include/unity_config.h:1]: Why is it empty?

File: test/test_indicator/test_indicator.c (hunk 4 above)
Comment (Tonidotpy) [reply test/test_indicator/test_indicator.c:34]: All the other fields in the handler should also be checked if they are initialized correctly for this test.
The `MEMORY` variant of the assert can be used, check the unity documentation for more info.

File: test/test_indicator/test_indicator.c
Comment (Tonidotpy) [reply test/test_indicator/test_indicator.c:1]: I think it is the `test_indicator_initialization_state` function even if it is missing the call to the actual init function.
Test for invalid input is still missing such as if the function pointer given as parameter is `NULL`.

File: test/test_indicator/test_indicator.c (hunk 3 above)
Comment (Tonidotpy) [reply test/test_indicator/test_indicator.c:58]: Yes and no, I don't think testing the `indicators_set` function is really needed since it is only just a variable assignment.
There is one test that is missing which is for invalid input, such as if i pass to the function an `indicator` enum value which is out of range, this test should fail since there is no such check for that.

File: test/test_feedback/test_feedback.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    feedback_update_state();
+
+    // Verify mock was called correct number of times (once per indicator)
+    TEST_ASSERT_EQUAL_INT_MESSAGE(
+        FEEDBACK_NAME_COUNT,
+        fake_read_feedback_fake.call_count,
+        "Update should poll all sensors");
+}
+
+void test_feedback_get_state(void) {
```
Comment (Tonidotpy) [inline test/test_feedback/test_feedback.c:38]: There is no need to check the `feedback_get_state` function with valid input becuase it just returns a value.
What is missing here is the check for invalid input, if I give an invalid `feedback` enum value the test should fail since there is no such check right now.

File: test/test_feedback/test_feedback.c
Comment (Tonidotpy) [reply test/test_feedback/test_feedback.c:1]: I agree that `feedback_init` is not tested but `feedback_get_state` only require test for invalid input in my opinion.

**Review by Tonidotpy — CHANGES_REQUESTED** [review]
I would like to review the source code as well as the tests in person with @Bridiro and @MattWaX to change a couple of things to simplify the development and maintainance of the code.

File: test/test_feedback/test_feedback.c
Comment (MattWaX) [reply test/test_feedback/test_feedback.c:1]: For `feedback_init` should I create a new function to return the init state, check directly from the handler or just call `feedback_init` again and check for the return code?
